wgengine/magicsock/magicsock.go

```diff
@@ -2686,16 +2694,50 @@ type discoInfo struct {
 type endpointTrackerEntry struct {
 	endpoint tailcfg.Endpoint
 	until    time.Time
+	index    int
+}
+
+type endpointHeap []*endpointTrackerEntry
+
+var _ heap.Interface[*endpointTrackerEntry] = (*endpointHeap)(nil)
+
+func (eh endpointHeap) Len() int { return len(eh) }
+
+func (eh endpointHeap) Less(i, j int) bool {
+	// We want to store items so that the lowest item in the heap is the
+	// oldest, so that heap.Pop()-ing from the endpointHeap will remove the
+	// oldest entry.
+	return eh[i].until.Before(eh[j].until)
+}
+
+func (eh endpointHeap) Swap(i, j int) {
+	eh[i], eh[j] = eh[j], eh[i]
+	eh[i].index = i
+	eh[j].index = j
+}
+
+func (eh *endpointHeap) Push(item *endpointTrackerEntry) {
+	n := len(*eh)
+	item.index = n
+	*eh = append(*eh, item)
+}
+
+func (eh *endpointHeap) Pop() *endpointTrackerEntry {
+	old := *eh
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	item.index = -1 // for safety
+	*eh = old[0 : n-1]
+	return item
 }
 
 type endpointTracker struct {
-	mu    sync.Mutex
-	cache map[netip.AddrPort]endpointTrackerEntry
+	mu        sync.Mutex
+	endpoints map[netip.Addr]endpointHeap
 }
```

Comment:
while you're rewriting this anyway, now would be a good time to move this all to its own file in this directory.

also, can you get some more (any) docs on these three types?

Looks good at first glance but I'm reviewing it more closely now.

Reply:
Added docs, moved to a new file, moved tests to a new file as well

Reply:
Side benefit of pulling this out to a new file: I confirmed that there's 100% test coverage, other than the "defense in depth" / "should never happen" comment 🎉